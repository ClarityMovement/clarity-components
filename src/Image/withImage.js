import T from 'prop-types';
import compose from 'recompose/compose';
import withReducer from 'recompose/withReducer';
import mapProps from 'recompose/mapProps';
import pure from 'recompose/pure';
import flattenProp from 'recompose/flattenProp';

import omitProps from '../hocs/omitProps';
import omitPropTypes from '../hocs/omitPropTypes';
import extendStatics from '../hocs/extendStatics';
import copyStatics from '../hocs/copyStatics';
import embedHandlers from '../hocs/embedHandlers';

const defaultState = {
  scale: 1,
  uploading: false,
  failed: false,
  editor: undefined,
};

const createInitialState = ({ initialState, ...props }) => {
  const state = {
    ...defaultState,
    ...initialState,
  };

  Object.keys(defaultState).forEach((key) => {
    if (typeof props[key] !== 'undefined') {
      state[key] = props[key];
    }
  });

  return state;
};

const onUpload = ({
  editor,
  uploadImage,
  onUploadFail,
  onUploadStart,
  onUploadSucceed,
}) => () => {
  if (!uploadImage) return;

  const dataUrl = editor.getDataUrl();
  onUploadStart();
  uploadImage(dataUrl).then(onUploadSucceed, onUploadFail);
};

const reset = ({ setState, ...props }) => () => setState({ type: 'reset', props });
const setEditor = ({ setState }) => editor => setState({ editor });
const openEditor = ({ setState }) => () => setState({ uploaded: false });
const setImage = ({ setState }) => image => setState({ image });
const setScale = ({ setState }) => scale => setState({ scale });
const onUploadSucceed = ({ setState }) => ({ url }) =>
  setState({ url, uploaded: true, uploading: false });
const onUploadFail = ({ setState }) => () =>
  setState({ failed: true, uploading: false });
const onUploadStart = ({ setState }) => () =>
  setState({ uploading: true, uploaded: false, failed: false });

export const handlers = [{
  openEditor,
  setEditor,
  setImage,
  setScale,
  onUploadFail,
  onUploadStart,
  onUploadSucceed,
  reset,
}, {
  onUpload,
}];

const mergeState = (state, { type, props, ...action }) => {
  if (type === 'reset') {
    const { editor } = state;
    if (!props.image && editor && editor.reset) {
      editor.reset();
    }
    return createInitialState({ ...props, editor });
  }
  return { ...state, ...action };
};

const propsMapper = ({
  url,
  uploaded = !!url,
  image = (url ? { preview: url } : undefined),
  ...rest
}) => ({
  ...rest,
  url,
  uploaded,
  image,
  cropping: !!image && !uploaded,
});

export const propTypes = {
  initialState: T.shape({
    url: T.string,
  }),
  onUploadFail: T.func,
  onUploadStart: T.func,
  onUploadSucceed: T.func,
  uploadImage: T.func,
};

export const defaultProps = {
  initialState: defaultState,
};

export default Component => compose(
  omitPropTypes(['onUpload']),
  extendStatics({
    displayName: 'withImage',
    propTypes,
    defaultProps,
  }),
  copyStatics(Component),
  withReducer('state', 'setState', mergeState, createInitialState),
  flattenProp('state'),
  omitProps('initialState'),
  embedHandlers(handlers),
  omitProps(['editor', 'state', 'setState']),
  mapProps(propsMapper),
  pure,
)(Component);
