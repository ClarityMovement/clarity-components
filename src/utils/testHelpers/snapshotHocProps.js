import React from 'react';

import enzymeToJson from 'enzyme-to-json';
import { shallowWithUntil as shallow } from './until';

const snapshotPropsMapperHoc = (hoc, inputProps) => {
  // This Component can't render deeply, must work with shallow renderer.
  const PropsDisplayer = outputProps => (
    <div name="props-displayer">
      <input-props {...inputProps} />
      <output-props {...outputProps} />
    </div>
  );
  const NewComponent = hoc(PropsDisplayer);
  const wrapper = shallow(<NewComponent {...inputProps} />).until(PropsDisplayer);
  expect(enzymeToJson(wrapper)).toMatchSnapshot();
};

export default snapshotPropsMapperHoc;
