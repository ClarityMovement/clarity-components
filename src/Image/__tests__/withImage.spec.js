/* eslint-env jest */
import snapshotHocProps from '../../utils/testHelpers/snapshotHocProps';
import withImage from '../withImage';

const url = 'http://pic.example.com/200/200';
const image = { preview: url };

describe('withImage(BaseComponent): NewComponent', () => {
  let props;

  afterEach(() => snapshotHocProps(withImage, props));

  describe('start with default props', () => {
    test('waiting for user to select any image', () => {
      props = {};
    });
  });

  describe('start with image preset', () => {
    test('cropping image', () => {
      props = { image };
    });
    test('uploading image', () => {
      props = { image, uploading: true };
    });
    test('failed to upload image', () => {
      props = { image, failed: true };
    });
    test('succeeded to upload image', () => {
      props = { image, url };
    });
  });

  describe('start with url preset', () => {
    test('previewing url', () => {
      props = { url };
    });
    test('cropping url', () => {
      props = { url, uploaded: false };
    });
  });
});
