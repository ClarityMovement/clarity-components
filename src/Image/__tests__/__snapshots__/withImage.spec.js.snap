// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withImage(BaseComponent): NewComponent start with default props waiting for user to select any image 1`] = `
<div
  name="props-displayer"
>
  <input-props />
  <output-props
    cropping={false}
    failed={false}
    onUpload={[Function]}
    onUploadFail={[Function]}
    onUploadStart={[Function]}
    onUploadSucceed={[Function]}
    openEditor={[Function]}
    reset={[Function]}
    setEditor={[Function]}
    setImage={[Function]}
    uploaded={false}
    uploading={false}
  />
</div>
`;

exports[`withImage(BaseComponent): NewComponent start with image preset cropping image 1`] = `
<div
  name="props-displayer"
>
  <input-props
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
  />
  <output-props
    cropping={true}
    failed={false}
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    onUpload={[Function]}
    onUploadFail={[Function]}
    onUploadStart={[Function]}
    onUploadSucceed={[Function]}
    openEditor={[Function]}
    reset={[Function]}
    setEditor={[Function]}
    setImage={[Function]}
    uploaded={false}
    uploading={false}
  />
</div>
`;

exports[`withImage(BaseComponent): NewComponent start with image preset failed to upload image 1`] = `
<div
  name="props-displayer"
>
  <input-props
    failed={true}
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
  />
  <output-props
    cropping={true}
    failed={true}
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    onUpload={[Function]}
    onUploadFail={[Function]}
    onUploadStart={[Function]}
    onUploadSucceed={[Function]}
    openEditor={[Function]}
    reset={[Function]}
    setEditor={[Function]}
    setImage={[Function]}
    uploaded={false}
    uploading={false}
  />
</div>
`;

exports[`withImage(BaseComponent): NewComponent start with image preset succeeded to upload image 1`] = `
<div
  name="props-displayer"
>
  <input-props
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    url="http://pic.example.com/200/200"
  />
  <output-props
    cropping={false}
    failed={false}
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    onUpload={[Function]}
    onUploadFail={[Function]}
    onUploadStart={[Function]}
    onUploadSucceed={[Function]}
    openEditor={[Function]}
    reset={[Function]}
    setEditor={[Function]}
    setImage={[Function]}
    uploaded={true}
    uploading={false}
    url="http://pic.example.com/200/200"
  />
</div>
`;

exports[`withImage(BaseComponent): NewComponent start with image preset uploading image 1`] = `
<div
  name="props-displayer"
>
  <input-props
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    uploading={true}
  />
  <output-props
    cropping={true}
    failed={false}
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    onUpload={[Function]}
    onUploadFail={[Function]}
    onUploadStart={[Function]}
    onUploadSucceed={[Function]}
    openEditor={[Function]}
    reset={[Function]}
    setEditor={[Function]}
    setImage={[Function]}
    uploaded={false}
    uploading={true}
  />
</div>
`;

exports[`withImage(BaseComponent): NewComponent start with url preset cropping url 1`] = `
<div
  name="props-displayer"
>
  <input-props
    uploaded={false}
    url="http://pic.example.com/200/200"
  />
  <output-props
    cropping={true}
    failed={false}
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    onUpload={[Function]}
    onUploadFail={[Function]}
    onUploadStart={[Function]}
    onUploadSucceed={[Function]}
    openEditor={[Function]}
    reset={[Function]}
    setEditor={[Function]}
    setImage={[Function]}
    uploaded={false}
    uploading={false}
    url="http://pic.example.com/200/200"
  />
</div>
`;

exports[`withImage(BaseComponent): NewComponent start with url preset previewing url 1`] = `
<div
  name="props-displayer"
>
  <input-props
    url="http://pic.example.com/200/200"
  />
  <output-props
    cropping={false}
    failed={false}
    image={
      Object {
        "preview": "http://pic.example.com/200/200",
      }
    }
    onUpload={[Function]}
    onUploadFail={[Function]}
    onUploadStart={[Function]}
    onUploadSucceed={[Function]}
    openEditor={[Function]}
    reset={[Function]}
    setEditor={[Function]}
    setImage={[Function]}
    uploaded={true}
    uploading={false}
    url="http://pic.example.com/200/200"
  />
</div>
`;
